src/renderers/shaders/ShaderChunk/normal_fragment_begin.glsl.js

```diff
@@ -13,25 +13,26 @@ float faceDirection = gl_FrontFacing ? 1.0 : - 1.0;
 
 	#ifdef DOUBLE_SIDED
 
-		normal = normal * faceDirection;
+		normal *= faceDirection;
 
 	#endif
 
-	#ifdef USE_TANGENT
+	#if defined( TANGENTSPACE_NORMALMAP ) || defined( USE_CLEARCOAT_NORMALMAP )
 
-		vec3 tangent = normalize( vTangent );
-		vec3 bitangent = normalize( vBitangent );
+		#ifdef USE_TANGENT
 
-		#ifdef DOUBLE_SIDED
+			mat3 tbn = mat3( normalize( vTangent ), normalize( vBitangent ), normal );
+
+		#else
 
-			tangent = tangent * faceDirection;
-			bitangent = bitangent * faceDirection;
+			mat3 tbn = getTangentSpace( - vViewPosition, normal );
```

Comment:
So that we can have a proper thread:

> If flatShading = true , it appears the TBN matrix is not defined.

I believe this was true before too, as it's not used in that case. Although, I suppose it depends on what flat shading is intended to mean? Does it just mean that interpolated normals are not applied but all the rest (normal maps, anisotropy, clearcoat normal maps) are? I believe currently it disables everything normal-related.

Reply:
Nevermind, I see it now. Will fix.

Reply:
1. The shader is still throwing errors for me if `flatShading` is true.

2. When a normal map is present, the three.js convention, adhered to years ago, is for the back face to render the same, whether `side` is `frontSide` or `doubleSide`.

Reply:
1. The shader is still throwing errors for me if `flatShading` is true and tangents and a normal map are present.

2. When a normal map is present, the three.js convention, adhered to years ago, is for the back face to render the same, whether `side` is `BackSide` or `DoubleSide`.

Reply:
> The `side` was not applied to the normal before for flat shading, which seems like a bug, so that's corrected here. Please let me know if that's an incorrect assumption. 

Better to make this PR only about refactoring with no side effects and make a new PR for fixing bugs.

Reply:
Okay, will fix. @WestLangley what errors are you getting? Might you happen to have a GLB with tangents, a normal map and no normals so I can repro?

Reply:
I have invested considerable time trying to determine what is working, what bugs are preexisting, and what bugs are introduced in this PR.

Everything was working as intended in r125. (2 years ago).

>When a normal map is present, the three.js convention, adhered to years ago, is for the back face to render the same, whether side is BackSide or DoubleSide.

In this PR, a back face normal map is rendered incorrectly when `material.side = BackSide` and tangents are present.

I believe this bug was introduced in #22146, as mentioned in https://github.com/mrdoob/three.js/pull/25580#issuecomment-1464746778.

>The shader is still throwing errors for me if flatShading is true and tangents and a normal map are present.

I believe this is a preexisting bug, too.

//

https://github.com/cx20/gltf-test/tree/master/tutorialModels/NormalTangentTest is one testbed. You can remove the normal attribute in the loader callback. Or you can just set `flatShading = true`.

Reply:
Great, thank you. I thought you meant the shader wasn't compiling, but if it's just the double-sided change messing with rendering then I'll get that reverted.

Reply:
Okay, I think I have it fixed.

Reply:
Let me know if you opt to fix (1) or (2) in my previous comment, and I will recheck.

Reply:
> I believe this bug was introduced in https://github.com/mrdoob/three.js/pull/22146, as mentioned in https://github.com/mrdoob/three.js/pull/25580#issuecomment-1464746778.

@donmccurdy 🤔

Reply:
> I believe this bug was introduced in https://github.com/mrdoob/three.js/pull/22146, as mentioned in https://github.com/mrdoob/three.js/pull/25580#issuecomment-1464746778.

Filed to https://github.com/mrdoob/three.js/issues/25701. 

> The shader is not compiling for me if flatShading is true and tangents and a normal map are present.

What should happen here? I'm inclined to say combining `.flatShading` and `.normalMap` is not allowed, and that a normal map requires vertex normals. Or do we need to support that combination?